File: samples/bot-cards/csharp/TeamsFileUpload/Bots/TeamsFileUploadBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Teams;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json.Linq;

namespace Microsoft.BotBuilderSamples.Bots
{
    public class TeamsFileUploadBot : TeamsActivityHandler
    {
        // Private fields for the HttpClient and bot credentials.
        private readonly IHttpClientFactory _clientFactory;
        private static HttpClient _client;
        private static string microsoftAppId;
        private static string microsoftAppPassword;

        // Constructor that initializes the bot with the configuration and client factory.
        public TeamsFileUploadBot(IHttpClientFactory clientFactory, IConfiguration configuration)
        {
            _clientFactory = clientFactory;

            // Initialize static variables only once.
            if (_client == null)
            {
                _client = _clientFactory.CreateClient();
            }

            microsoftAppId = configuration["MicrosoftAppId"];
            microsoftAppPassword = configuration["MicrosoftAppPassword"];
        }

        // Handles incoming message activities and processes file uploads or inline images.
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            bool messageWithFileDownloadInfo = turnContext.Activity.Attachments?[0].ContentType == FileDownloadInfo.ContentType;

            if (messageWithFileDownloadInfo)
            {
                // Process file download.
                var file = turnContext.Activity.Attachments[0];
                var fileDownload = JObject.FromObject(file.Content).ToObject<FileDownloadInfo>();
                string filePath = Path.Combine("Files", file.Name);

                var response = await _client.GetAsync(fileDownload.DownloadUrl);
                await SaveFileAsync(filePath, response);

                var reply = MessageFactory.Text($"<b>{file.Name}</b> received and saved.");
                reply.TextFormat = "xml";
                await turnContext.SendActivityAsync(reply, cancellationToken);
            }
            else if (turnContext.Activity.Attachments?[0].ContentType.Contains("image/*") == true)
            {
                // Process inline image.
                await ProcessInlineImage(turnContext, cancellationToken);
            }
            else
            {
                // Send a default file.
                string filename = "teams-logo.png";
                string filePath = Path.Combine("Files", filename);
                long fileSize = new FileInfo(filePath).Length;
                await SendFileCardAsync(turnContext, filename, fileSize, cancellationToken);
            }
        }

        // Helper method to save file to disk.
        private async Task SaveFileAsync(string filePath, HttpResponseMessage response)
        {
            using (var fileStream = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.None))
            {
                await response.Content.CopyToAsync(fileStream);
            }
        }

        // Processes inline image attachments and saves them.
        private async Task ProcessInlineImage(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            var attachment = turnContext.Activity.Attachments[0];
            var token = await new MicrosoftAppCredentials(microsoftAppId, microsoftAppPassword).GetTokenAsync();
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);

            var responseMessage = await _client.GetAsync(attachment.ContentUrl);

            // Save the inline image to Files directory.
            var filePath = Path.Combine("Files", "ImageFromUser.png");
            await SaveFileAsync(filePath, responseMessage);

            // Create reply with image.
            var reply = MessageFactory.Text($"Attachment of {attachment.ContentType} type and size of {responseMessage.Content.Headers.ContentLength} bytes received.");
            reply.Attachments = new List<Attachment> { GetInlineAttachment() };
            await turnContext.SendActivityAsync(reply, cancellationToken);
        }

        // Creates an inline attachment for the image that was received.
        private static Attachment GetInlineAttachment()
        {
            var imagePath = Path.Combine("Files", "ImageFromUser.png");
            var imageData = Convert.ToBase64String(File.ReadAllBytes(imagePath));

            return new Attachment
            {
                Name = @"ImageFromUser.png",
                ContentType = "image/png",
                ContentUrl = $"data:image/png;base64,{imageData}",
            };
        }

        // Sends a file consent card to the user.
        private async Task SendFileCardAsync(ITurnContext turnContext, string filename, long filesize, CancellationToken cancellationToken)
        {
            var consentContext = new Dictionary<string, string>
            {
                { "filename", filename },
            };

            var fileCard = new FileConsentCard
            {
                Description = "This is the file I want to send you",
                SizeInBytes = filesize,
                AcceptContext = consentContext,
                DeclineContext = consentContext,
            };

            var asAttachment = new Attachment
            {
                Content = fileCard,
                ContentType = FileConsentCard.ContentType,
                Name = filename,
            };

            var replyActivity = turnContext.Activity.CreateReply();
            replyActivity.Attachments = new List<Attachment> { asAttachment };
            await turnContext.SendActivityAsync(replyActivity, cancellationToken);
        }

        // Handles file consent acceptance.
        protected override async Task OnTeamsFileConsentAcceptAsync(ITurnContext<IInvokeActivity> turnContext, FileConsentCardResponse fileConsentCardResponse, CancellationToken cancellationToken)
        {
            try
            {
                JToken context = JObject.FromObject(fileConsentCardResponse.Context);
                string filePath = Path.Combine("Files", context["filename"].ToString());
                long fileSize = new FileInfo(filePath).Length;

                using (var fileStream = File.OpenRead(filePath))
                {
                    var fileContent = new StreamContent(fileStream);
                    fileContent.Headers.ContentLength = fileSize;
                    fileContent.Headers.ContentRange = new ContentRangeHeaderValue(0, fileSize - 1, fileSize);
                    await _client.PutAsync(fileConsentCardResponse.UploadInfo.UploadUrl, fileContent, cancellationToken);
                }

                await FileUploadCompletedAsync(turnContext, fileConsentCardResponse, cancellationToken);
            }
            catch (Exception e)
            {
                await FileUploadFailedAsync(turnContext, e.ToString(), cancellationToken);
            }
        }

        // Handles file consent decline.
        protected override async Task OnTeamsFileConsentDeclineAsync(ITurnContext<IInvokeActivity> turnContext, FileConsentCardResponse fileConsentCardResponse, CancellationToken cancellationToken)
        {
            JToken context = JObject.FromObject(fileConsentCardResponse.Context);
            var reply = MessageFactory.Text($"Declined. We won't upload file <b>{context["filename"]}</b>.");
            reply.TextFormat = "xml";
            await turnContext.SendActivityAsync(reply, cancellationToken);
        }

        // Sends a reply indicating that file upload is completed.
        private async Task FileUploadCompletedAsync(ITurnContext turnContext, FileConsentCardResponse fileConsentCardResponse, CancellationToken cancellationToken)
        {
            var downloadCard = new FileInfoCard
            {
                UniqueId = fileConsentCardResponse.UploadInfo.UniqueId,
                FileType = fileConsentCardResponse.UploadInfo.FileType,
            };

            var asAttachment = new Attachment
            {
                Content = downloadCard,
                ContentType = FileInfoCard.ContentType,
                Name = fileConsentCardResponse.UploadInfo.Name,
                ContentUrl = fileConsentCardResponse.UploadInfo.ContentUrl,
            };

            var reply = MessageFactory.Text($"<b>File uploaded.</b> Your file <b>{fileConsentCardResponse.UploadInfo.Name}</b> is ready to download");
            reply.TextFormat = "xml";
            reply.Attachments = new List<Attachment> { asAttachment };

            await turnContext.SendActivityAsync(reply, cancellationToken);
        }

        // Sends a reply indicating that file upload has failed.
        private async Task FileUploadFailedAsync(ITurnContext turnContext, string error, CancellationToken cancellationToken)
        {
            var reply = MessageFactory.Text($"<b>File upload failed.</b> Error: <pre>{error}</pre>");
            reply.TextFormat = "xml";
            await turnContext.SendActivityAsync(reply, cancellationToken);
        }
    }
}

File: samples/bot-file-upload/nodejs/bots/teamsFileUploadBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const fs = require('fs');
const path = require('path');
const axios = require('axios');
const { TurnContext, MessageFactory, TeamsActivityHandler } = require('botbuilder');
const { MicrosoftAppCredentials } = require('botframework-connector');
const { generateFileName, getFileSize, writeFile } = require('../services/fileService');

const FILES_DIR = 'files';

/**
 * TeamsFileUploadBot class handles file upload interactions with Teams.
 */
class TeamsFileUploadBot extends TeamsActivityHandler {
    constructor() {
        super();

        this.onMessage(async (context, next) => {
            TurnContext.removeRecipientMention(context.activity);
            const attachment = context.activity.attachments && context.activity.attachments[0];

            if (attachment) {
                const imageRegex = /image\/.*/;

                if (attachment.contentType === 'application/vnd.microsoft.teams.file.download.info') {
                    // Handle file download info
                    await this.handleFileDownload(attachment, context);
                } else if (imageRegex.test(attachment.contentType)) {
                    // Handle inline image
                    await this.processInlineImage(context);
                } else {
                    // Send a default file card if no relevant attachment is found
                    await this.sendFileCard(context);
                }
            }
            await next();
        });
    }

    /**
     * Handles file download and saves the file.
     * @param {object} file - The file attachment.
     * @param {object} context - The bot context.
     */
    async handleFileDownload(file, context) {
        const config = {
            responseType: 'stream'
        };
        const filePath = path.join(FILES_DIR, file.name);
        await writeFile(file.content.downloadUrl, config, filePath);

        const reply = MessageFactory.text(`<b>${file.name}</b> received and saved.`);
        reply.textFormat = 'xml';
        await context.sendActivity(reply);
    }

    /**
     * Sends a file card for user consent.
     * @param {object} context - The bot context.
     */
    async sendFileCard(context) {
        const filename = 'teams-logo.png';
        const stats = fs.statSync(path.join(FILES_DIR, filename));
        const fileSize = stats.size;

        const consentContext = { filename: filename };
        const fileCard = {
            description: 'This is the file I want to send you',
            sizeInBytes: fileSize,
            acceptContext: consentContext,
            declineContext: consentContext
        };

        const asAttachment = {
            content: fileCard,
            contentType: 'application/vnd.microsoft.teams.card.file.consent',
            name: filename
        };
        await context.sendActivity({ attachments: [asAttachment] });
    }

    /**
     * Handles file consent acceptance and uploads the file.
     * @param {object} context - The bot context.
     * @param {object} fileConsentCardResponse - The consent response.
     */
    async handleTeamsFileConsentAccept(context, fileConsentCardResponse) {
        try {
            const fname = path.join(FILES_DIR, fileConsentCardResponse.context.filename);
            const fileInfo = fs.statSync(fname);
            const fileContent = fs.createReadStream(fname); // Use stream instead of reading into memory

            await axios.put(
                fileConsentCardResponse.uploadInfo.uploadUrl,
                fileContent, {
                    headers: {
                        'Content-Type': 'image/png',
                        'Content-Length': fileInfo.size,
                        'Content-Range': `bytes 0-${fileInfo.size - 1}/${fileInfo.size}`
                    }
                });

            await this.fileUploadCompleted(context, fileConsentCardResponse);
        } catch (e) {
            await this.fileUploadFailed(context, e.message);
        }
    }

    /**
     * Handles file consent decline and informs the user.
     * @param {object} context - The bot context.
     * @param {object} fileConsentCardResponse - The consent response.
     */
    async handleTeamsFileConsentDecline(context, fileConsentCardResponse) {
        const reply = MessageFactory.text(`The file <b>${fileConsentCardResponse.context.filename}</b> has been declined and will not be uploaded.`);
        reply.textFormat = 'xml';
        await context.sendActivity(reply);
    }

    /**
     * Notifies the user when the file upload is completed.
     * @param {object} context - The bot context.
     * @param {object} fileConsentCardResponse - The consent response.
     */
    async fileUploadCompleted(context, fileConsentCardResponse) {
        const downloadCard = {
            uniqueId: fileConsentCardResponse.uploadInfo.uniqueId,
            fileType: fileConsentCardResponse.uploadInfo.fileType
        };

        const asAttachment = {
            content: downloadCard,
            contentType: 'application/vnd.microsoft.teams.card.file.info',
            name: fileConsentCardResponse.uploadInfo.name,
            contentUrl: fileConsentCardResponse.uploadInfo.contentUrl
        };

        const reply = MessageFactory.text(`<b>Your file ${fileConsentCardResponse.uploadInfo.name}</b> has been successfully uploaded and is ready to download.`);
        reply.textFormat = 'xml';
        reply.attachments = [asAttachment];
        await context.sendActivity(reply);
    }

    /**
     * Handles failed file upload and notifies the user.
     * @param {object} context - The bot context.
     * @param {string} error - The error message.
     */
    async fileUploadFailed(context, error) {
        const reply = MessageFactory.text(`<b>File upload failed.</b> Error: <pre>${error}</pre>`);
        reply.textFormat = 'xml';
        await context.sendActivity(reply);
    }

    /**
     * Processes an inline image by saving it and notifying the user.
     * @param {object} context - The bot context.
     */
    async processInlineImage(context) {
        const file = context.activity.attachments[0];
        const credentials = new MicrosoftAppCredentials(process.env.MicrosoftAppId, process.env.MicrosoftAppPassword);
        const botToken = await credentials.getToken();

        const config = {
            headers: { Authorization: `Bearer ${botToken}` },
            responseType: 'stream'
        };

        const fileName = await generateFileName(FILES_DIR);
        const filePath = path.join(FILES_DIR, fileName);
        await writeFile(file.contentUrl, config, filePath);

        const fileSize = await getFileSize(filePath);
        const reply = MessageFactory.text(`Image <b>${fileName}</b> of size <b>${fileSize}</b> bytes received and saved.`);
        const inlineAttachment = this.getInlineAttachment(fileName);

        reply.attachments = [inlineAttachment];
        await context.sendActivity(reply);
    }

    /**
     * Creates an inline attachment for the image.
     * @param {string} fileName - The name of the file.
     * @returns {object} The inline attachment object.
     */
    getInlineAttachment(fileName) {
        const imageData = fs.readFileSync(path.join(FILES_DIR, fileName));
        const base64Image = Buffer.from(imageData).toString('base64');

        return {
            name: fileName,
            contentType: 'image/png',
            contentUrl: `data:image/png;base64,${base64Image}`
        };
    }
}

module.exports.TeamsFileUploadBot = TeamsFileUploadBot;
